opt_einsum/paths.py

```diff
@@ -193,154 +220,230 @@ def _optimal_iterate(path, remaining, inputs, flops):
+                    if new_flops < self.best['flops']:
+                        self.best['flops'] = new_flops
+                        self.best['ssa_path'] = path + (tuple(remaining),)
+                    return None
+
+                # set cost heuristic in order to locally sort possible contractions
+                size1, size2 = size_cache[inputs[i]], size_cache[inputs[j]]
+                cost = self.cost_fn(size12, size1, size2, k12, k1, k2)
 
-            candidate = _assess_candidate(k1, k2, i, j)
-            if candidate:
-                heapq.heappush(candidates, candidate)
+                return cost, flops12, new_flops, new_size, (i, j), k12
 
-        # assess outer products if nothing left
-        if not candidates:
+            # check all possible remaining paths
+            candidates = []
             for i, j in itertools.combinations(remaining, 2):
+                if i > j:
+                    i, j = j, i
                 k1, k2 = inputs[i], inputs[j]
+
+                # initially ignore outer products
+                if k1.isdisjoint(k2):
+                    continue
+
                 candidate = _assess_candidate(k1, k2, i, j)
                 if candidate:
                     heapq.heappush(candidates, candidate)
 
-        if not candidates:
-            return
-
-        # recurse into all or some of the best candidate contractions
-        bi = 0
-        while (nbranch is None or bi < nbranch) and candidates:
-            _, _, new_flops, (i, j), k12 = heapq.heappop(candidates)
-            _branch_iterate(path=path + ((i, j),),
-                            inputs=inputs + (k12,),
-                            remaining=remaining - {i, j} | {len(inputs)},
-                            flops=new_flops)
-            bi += 1
-
-    _branch_iterate(path=(),
-                    inputs=inputs,
-                    remaining=set(range(len(inputs))),
-                    flops=0)
-
-    return ssa_to_linear(best['path'])
-
-
-def _get_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2):
+            # assess outer products if nothing left
+            if not candidates:
+                for i, j in itertools.combinations(remaining, 2):
+                    if i > j:
+                        i, j = j, i
+                    k1, k2 = inputs[i], inputs[j]
+                    candidate = _assess_candidate(k1, k2, i, j)
+                    if candidate:
+                        heapq.heappush(candidates, candidate)
+
+            # recurse into all or some of the best candidate contractions
+            bi = 0
+            while (self.nbranch is None or bi < self.nbranch) and candidates:
+                _, _, new_flops, new_size, (i, j), k12 = heapq.heappop(candidates)
+                _branch_iterate(path=path + ((i, j),),
+                                inputs=inputs + (k12,),
+                                remaining=remaining - {i, j} | {len(inputs)},
```

Comment:
Do we need to worry slightly about order of operations here? I try to be fairly specific just so it is very clear to the user.

Reply:
Not sure I quite understand - could you elaborate?

Reply:
`remaining - ({i, j} | {len(inputs)})`

Reply:
Ah I see, I'll add some parentheses.